File: app/src/main/java/com/mthaler/knittings/KnittingListFragment.java
package com.mthaler.knittings;


import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.ListFragment;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;
import java.text.DateFormat;
import java.util.ArrayList;
import java.util.Date;

public class KnittingListFragment extends ListFragment implements KnittingListView {

    private ArrayList<Knitting> knittings;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getActivity().setTitle(R.string.knittings_title);
        this.knittings = KnittingsDataSource.getInstance(getActivity()).getAllKnittings();
        KnittingAdapter adapter = new KnittingAdapter(knittings);
        setListAdapter(adapter);
    }

    @Override
    public void onListItemClick(ListView l, View v, int position, long id) {
        // get the Crime from the adapter
        Knitting c = ((KnittingAdapter)getListAdapter()).getItem(position);
        // start an instance of CrimePagerActivity
        Intent i = new Intent(getActivity(), KnittingActivity.class);
        i.putExtra(KnittingActivity.EXTRA_KNITTING_ID, c.getId());
        startActivityForResult(i, 0);
    }

    @Override
    public void onResume() {
        super.onResume();
        this.knittings = KnittingsDataSource.getInstance(getActivity()).getAllKnittings();
        KnittingAdapter adapter = new KnittingAdapter(knittings);
        setListAdapter(adapter);
    }

    public void addKnitting() {
        Knitting knitting = KnittingsDataSource.getInstance(getActivity()).createKnitting("", "", new Date(), null, 0.0, 0.0, 0.0);
        Intent intent = new Intent(getActivity(), KnittingActivity.class);
        intent.putExtra(KnittingActivity.EXTRA_KNITTING_ID, knitting.getId());
        startActivity(intent);
    }

    private class KnittingAdapter extends ArrayAdapter<Knitting> {

        public KnittingAdapter(ArrayList<Knitting> knittings) {
            super(getActivity(), android.R.layout.simple_list_item_1, knittings);
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            // if we weren't given a view, inflate one
            if (null == convertView) {
                convertView = getActivity().getLayoutInflater()
                        .inflate(R.layout.list_item_knitting, null);
            }

            // configure the view for this Crime
            final Knitting knitting = getItem(position);

            final TextView titleTextView = convertView.findViewById(R.id.knitting_list_item_titleTextView);
            titleTextView.setText(knitting.getTitle());

            final TextView descriptionTextView = convertView.findViewById(R.id.knitting_list_item_descriptionTextView);
            descriptionTextView.setText(knitting.getDescription());

            final TextView startedTextView = convertView.findViewById(R.id.knitting_list_item_startedTextView);
            startedTextView.setText(DateFormat.getDateInstance().format(knitting.getStarted()));

            if (knitting.getDefaultPhoto() != null && knitting.getDefaultPhoto().getPreview() != null) {
                final ImageView photoView = convertView.findViewById(R.id.knitting_list_item_photoImageView);
                photoView.setImageBitmap(knitting.getDefaultPhoto().getPreview());
            }

            return convertView;
        }
    }
}

File: app/src/main/kotlin/com/mthaler/knittings/KnittingListFragment.kt
package com.mthaler.knittings

import android.graphics.Color
import android.support.v4.app.ListFragment
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.ListView
import android.widget.TextView
import com.mthaler.knittings.database.datasource
import com.mthaler.knittings.details.EditKnittingDetailsActivity
import com.mthaler.knittings.details.KnittingDetailsActivity
import com.mthaler.knittings.model.Knitting
import org.jetbrains.anko.AnkoLogger
import java.text.DateFormat
import java.util.Date
import org.jetbrains.anko.support.v4.*

/**
 * KnittingListFragment displays a list of knittings
 */
class KnittingListFragment : ListFragment(), KnittingListView, AnkoLogger {

    private var _sorting: Sorting = Sorting.NewestFirst
    private var _filter: Filter = NoFilter

    override fun onListItemClick(l: ListView?, v: View?, position: Int, id: Long) {
        // get current knitting and start knitting details activitiy if it is not null
        (listAdapter as KnittingAdapter).getItem(position)?.let {
            startActivity<KnittingDetailsActivity>(KnittingDetailsActivity.EXTRA_KNITTING_ID to it.id)
        }
    }

    override fun onResume() {
        super.onResume()
        updateKnittingList()
    }

    override fun addKnitting() {
        // start knitting activity with newly created knitting
        val knitting = datasource.createKnitting("", "", Date(), null, 0.0, 0.0, 0.0)
        startActivity<EditKnittingDetailsActivity>(EditKnittingDetailsActivity.EXTRA_KNITTING_ID to knitting.id)
    }

    override fun updateKnittingList() {
        val knittings = datasource.allKnittings
        when(_sorting) {
            Sorting.NewestFirst -> knittings.sortByDescending { it.started}
            Sorting.OldestFirst -> knittings.sortBy { it.started }
            Sorting.Alphabetical -> knittings.sortBy { it.title.toLowerCase() }
        }
        val filtered = _filter.filter(knittings)
        val adapter = KnittingAdapter(filtered)
        listAdapter = adapter
    }

    override fun getSorting(): Sorting = _sorting

    override fun setSorting(sorting: Sorting) {
        _sorting = sorting
    }

    override fun getFilter(): Filter = _filter

    override fun setFilter(filter: Filter) {
        _filter = filter
    }


    private inner class KnittingAdapter(knittings: List<Knitting>) : ArrayAdapter<Knitting>(activity, android.R.layout.simple_list_item_1, knittings) {

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {

            val v = if (convertView != null) convertView else activity!!.layoutInflater.inflate(R.layout.list_item_knitting, parent, false)

            // configure the view for the knitting
            val knitting = getItem(position)

            val titleTextView = v.findViewById<TextView>(R.id.knitting_list_item_titleTextView)
            titleTextView.text = knitting.title

            val descriptionTextView = v.findViewById<TextView>(R.id.knitting_list_item_descriptionTextView)
            descriptionTextView.text = knitting.description

            val startedTextView = v.findViewById<TextView>(R.id.knitting_list_item_startedTextView)
            startedTextView.text = DateFormat.getDateInstance().format(knitting.started)

            val photoView = v.findViewById<ImageView>(R.id.knitting_list_item_photoImageView)
            // the list item views are reused, we always need to set bitmap, otherwise the previous bitmap is used
            if (knitting.defaultPhoto?.preview != null) {
                photoView.setImageBitmap(knitting.defaultPhoto.preview)
            } else {
                photoView.setImageBitmap(null)
            }

            val categoryIndicator = v.findViewById<CategoryIndicator>(R.id.knitting_list_item_categoryIndicator)
            if (knitting.category != null) {
                if (knitting.category.color != null) {
                    categoryIndicator.color = knitting.category.color
                } else {
                    categoryIndicator.color = Color.WHITE
                }
            } else {
                categoryIndicator.color = Color.WHITE
            }

            return v
        }
    }
}
